src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -691,6 +692,287 @@ bool Compiler::optJumpThread(BasicBlock* const block, BasicBlock* const domBlock
+                domIsSameRelop ? "the same" : "a reverse");
+
+        // If lcl is not tracked, assume we can't safely reason about interference
+        // or liveness.
+        //
+        const unsigned   prevTreeLcl    = prevTreeLHS->AsLclVarCommon()->GetLclNum();
+        LclVarDsc* const prevTreeLclDsc = lvaGetDesc(prevTreeLcl);
+
+        if (!prevTreeLclDsc->lvTracked)
+        {
+            JITDUMP(" -- prev tree defs untracked V%02u\n", prevTreeLcl);
+            break;
+        }
+
+        // If we've run out of room to keep track of defined locals, bail.
+        //
+        if (definedLocalsCount >= DEFINED_LOCALS_SIZE)
+        {
+            JITDUMP(" -- ran out of space for tracking kills\n");
+            break;
+        }
+
+        // See if we can safely move a copy of prevTreeRHS later, to replace tree.
+        // We can, if none of its lcls are killed.
+        //
+        definedLocals[definedLocalsCount++] = prevTreeLcl;
+
+        for (unsigned int i = 0; i < definedLocalsCount; i++)
+        {
+            if (gtHasRef(prevTreeRHS, definedLocals[i], /*def only*/ false))
+            {
+                JITDUMP(" -- prev tree ref to V%02u interferes\n", definedLocals[i]);
+                break;
+            }
+        }
+
+        // Heuristic: only forward sub a relop
+        //
+        if (!prevTreeRHS->OperIsCompare())
+        {
+            JITDUMP(" -- prev tree is not relop\n");
+            continue;
+        }
+
+        // Heuristic: if the lcl defined here is live out, forward sub
+        // won't result in the original becoming dead. If the local is not
+        // live out that still may happen, but it's less likely.
+        //
+        if (VarSetOps::IsMember(this, block->bbLiveOut, prevTreeLclDsc->lvVarIndex))
+        {
+            JITDUMP(" -- prev tree lcl V%02u is live-out\n", prevTreeLcl);
+            continue;
+        }
+
+        JITDUMP(" -- prev tree is viable candidate for relop fwd sub!\n");
+        candidateTree = prevTreeRHS;
+        reverse       = domIsRevRelop;
+    }
+
+    if (candidateTree == nullptr)
+    {
+        return false;
+    }
+
+    // Looks good -- we are going to forward-sub candidateTree
+    //
+    GenTree* const substituteTree = gtCloneExpr(candidateTree);
+
+    // If we need the reverse compare, make it so
+    //
+    if (reverse)
+    {
+        substituteTree->SetOper(GenTree::ReverseRelop(substituteTree->OperGet()));
+
+        // This substitute tree will have the VN the same as the original.
+        // (modulo excep sets...? hmmm)
+        substituteTree->SetVNsFromNode(tree);
+    }
+
+    // We just intentionally created a duplicate -- we don't want CSE to undo this.
+    // (hopefully, the original is now dead).
+    //
+    // Also this relop is now a subtree of a jump.
+    //
+    substituteTree->gtFlags |= (GTF_DONT_CSE | GTF_RELOP_JMP_USED);
+
+    gtReplaceTree(stmt, tree, substituteTree);
```

Comment:
Hmm, `gtReplaceTree` was dead code before this change and as such is sort of in the process of being deleted (in #59912)...

It appears here we have the use edge, so maybe just replace the operand and resequence the statement explicitly? It should be a bit cheaper I think.

Reply:
> maybe just replace the operand and resequence the statement explicitly?

Sure.